sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-generation.service.ts

```diff
@@ -0,0 +1,506 @@
+      .innerJoin("offering.institutionLocation", "institutionLocation")
+      .innerJoin("offering.educationProgram", "educationProgram")
+      .innerJoin("application.student", "student") // ! The student alias here is also used in sub query 'getExistsBlockRestrictionQuery'.
+      .innerJoin("student.user", "user")
+      .innerJoin("student.sinValidation", "sinValidation")
+      .innerJoin("application.msfaaNumber", "msfaaNumber")
+      .innerJoin("disbursement.disbursementValues", "disbursementValue")
+      .where("disbursement.dateSent is null")
+      .andWhere("disbursement.disbursementDate <= :disbursementMinDate", {
+        disbursementMinDate,
+      })
+      .andWhere("application.applicationStatus = :applicationStatus", {
+        applicationStatus: ApplicationStatus.completed,
+      })
+      .andWhere("msfaaNumber.dateSigned IS NOT NULL")
+      .andWhere("msfaaNumber.cancelledDate IS NULL")
+      .andWhere("sinValidation.isValidSIN = true")
+      .andWhere("offering.offeringIntensity = :offeringIntensity", {
+        offeringIntensity,
+      })
+      .andWhere(
+        `NOT EXISTS(${this.studentRestrictionService
+          .getExistsBlockRestrictionQuery()
+          .getSql()})`,
+      )
+      .andWhere(
+        "disbursement.disbursementScheduleStatus = :disbursementScheduleStatus",
+        {
+          disbursementScheduleStatus: DisbursementScheduleStatus.Pending,
+        },
+      )
+      .andWhere("disbursement.coeStatus = :coeStatus", {
+        coeStatus: COEStatus.completed,
+      })
+      .setParameter("restrictionActionType", stopFullTimeBCFunding)
+      // 'restrictionActions' parameter used inside sub-query.
+      .setParameter("restrictionActions", possibleRestrictionActions)
+      // Necessary, for instance, for the overawards processing.
+      .orderBy("disbursement.disbursementDate")
+      .getRawAndEntities();
+    return mapFromRawAndEntities<ECertDisbursementSchedule>(
+      queryResult,
+      "stopFullTimeBCFunding",
+    );
+  }
+
+  /**
+   * Get the overawards balance consolidated per student and per loan award to
+   * apply, if needed, the overawards deductions.
+   * @param disbursements all disbursements that are part of one e-Cert.
+   * @param entityManager used to execute the commands in the same transaction.
+   */
+  private async applyOverawardsDeductions(
+    disbursements: ECertDisbursementSchedule[],
+    entityManager: EntityManager,
+  ) {
+    const studentsIds = disbursements.map(
+      (disbursement) => disbursement.studentAssessment.application.student.id,
+    );
+    // Get all the overawards balances for the students that are part of the disbursements.
+    const overawardsBalance =
+      await this.disbursementOverawardService.getOverawardBalance(
+        studentsIds,
+        entityManager,
+      );
+    // Apply the overawards for every student, if needed.
+    for (const studentId of studentsIds) {
```

Comment:
Instead of iterating through studentIds, we can iterate through `Object.entries(overawardsBalance)` so that we can iterate through the primary data and also skip the if (line 258 - 261)

Reply:
Changed to a similar recommendation, please take a look.

Reply:
As we talked about, the provided example `[studentId, overAwardValue]` will result in `studentId` defined as `string` and `overAwardValue` defined as any.